design/machine-remediation-request.md

```diff
@@ -0,0 +1,142 @@
+      * [MachineRemediationRequest status API](#machineremediationrequest-status-api)
+    * [Risks and Mitigations](#risks-and-mitigations)
+  * [Design Details](#design-details)
+    * [Work Items](#work-items)
+    * [Dependencies](#dependencies)
+    * [Test Plan](#test-plan)
+    * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+    * [Version Skew Strategy](#version-skew-strategy)
+  * [Alternatives](#alternatives)
+  * [References](#references)
+
+<!-- Added by: dhellmann, at: Wed Jun 26 16:08:57 EDT 2019 -->
+
+<!--te-->
+
+## Summary
+
+This document explains the implementation of the Machine Remediation Request controller and
+how it will work together with Machine Health Check controller to restore unhealthy machines.
+
+## Motivation
+
+To provide a mechanism to apply remediation strategy on an unhealthy node, that will be different by type and by a provider.
+
+### Goals
+
+1. To provide the interface that will make possible to apply different remediation strategy for each cloud or bare metal providers.
+2. To provide details regarding the progress of the remediation operation.
+
+### Non-Goals
+
+1. To provide a remediation strategy for each cloud or bare metal providers.
+2. To provide the integration with the **MachineHealthCheck** controller.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+It's common that high available environment should have a mechanism to restore unhealthy nodes and when delete and create a new node looks feasible solution for a cloud provider, in case of bare metal hosts with additional infrastructure configuration it looks like too destructive approach.
+To make possible to specify the remediation strategy for different failure cases and for nodes that run on different providers, this proposal introduces **MachineRemediationRequest** controller that should monitor all **MachineRemediationRequest** CR's and apply the remediation strategy according to details specified under the CR.
+When the **MachineRemediationRequest** will apply the remediation strategy, one who is initiating it will be the **MachineHealthCheck** controller. It will monitor nodes, and for each unhealthy node will create **MachineRemediationRequest** CR, if after the specified timeout the **MachineRemediationRequest** will not have the `remediationState` equal to `Succeeded`, the **MachineHealthCheck** controller will notify about the remediation failure, and can try to use different type of the remediation strategy.
+
+#### MachineRemediationRequest spec API
+
+This proposal introduces a new API type: **MachineRemediationRequest**, that makes possible to create simple **MachineRemediationRequest** object with the type of the remediation strategy for the specific machine.
+
+##### Example
+
+```yaml
+apiVersion: machineremediationrequest.metal3.io/v1alpha1
+kind: MachineRemediationRequest
+metadata:
+  name: mrr-test
+  namespace: openshift-machine-api
+spec:
+  type: reboot
+  machineName: test-machine
```

Comment:
Do we need this to be a full object reference? Or at least to include an optional namespace field?

Reply:
i dont think machines can live outside of the openshift-machine-api namespace

Reply:
`MachineRemediationRequest` should be created in the machine namespace, so the name should be enough to get the relevant machine object

Reply:
That assertion about namespacing is enforced by OpenShift, but not by kubernetes or the upstream machine-api. I don't think it's something we want to back into this API if we're going to try to upstream it.

Reply:
I can change the namespace value to something not openshift specific, but I still think that this resource should be created under the machine namespace.